### PR TITLE
REF: Don't insert default type parameters by Implement Member actions

### LINKED ISSUE #6186: Unnecessary full generic substitution when implementing members
## Environment

* **IntelliJ Rust plugin version:** 0.3.131.3366-202
* **Rust toolchain version:** 1.48
* **IDE name and version:** CLion 2020.2
* **Operating system:** Windows 10

## Problem description

When using automatic implementation generation of trait functions, generic parameters have defaults explicitly provided, often with invalid imports.

## Steps to reproduce

```rust
trait Foo {
    fn foo() -> Result<HashMap<String, i16>, String>;
}

impl Foo for Bar {
    // auto-generate fn implmentation here and notice RandomState gets added
    // what's worse, a random import also gets added:
    // use tracing::stdlib::collections::hash_map::RandomState;
    fn foo() -> Result<HashMap<String, i16, RandomState>, String> {
    }
}
```

### PR DESCRIPTION
**Pull request by mchernyavsky**
Fixes #6186

Depends on https://github.com/intellij-rust/intellij-rust/pull/6948.

changelog: Don't insert default type parameters by `Implement Member` actions

### CODE REVIEW

**Comment by Undin** [comment]
@mchernyavsky could you resolve merge conflicts?

**Comment by mchernyavsky** [comment]
@Undin Done.

**Comment by Kobzol** [comment]
I wonder if `skipUnchangedDefaultTypeArguments` should just be true by default?

**Review by Undin — COMMENTED** [review]
The current implementation doesn't work with type aliases. For example, for the following code:
```rust
mod bar {
    use std::collections::hash_map::RandomState;

    pub(crate) type HashMap<T, K = i32, S = RandomState> = std::collections::HashMap<T, K, S>;
}

trait Foo {
    fn foo() -> bar::HashMap<String>;
}

impl Foo for i32 {
}
```

`Implement memebers` produces
```rust
use crate::bar::HashMap;

mod bar {
    use std::collections::hash_map::RandomState;

    pub(crate) type HashMap<T, K = i32, S = RandomState> = std::collections::HashMap<T, K, S>;
}

trait Foo {
    fn foo() -> bar::HashMap<String>;
}

impl Foo for i32 {
    fn foo() -> HashMap<String, i32, RandomState> {
        unimplemented!()
    }
}
```

although I expect
```rust
use crate::bar::HashMap;

mod bar {
    use std::collections::hash_map::RandomState;

    pub(crate) type HashMap<T, K = i32, S = RandomState> = std::collections::HashMap<T, K, S>;
}

trait Foo {
    fn foo() -> bar::HashMap<String>;
}

impl Foo for i32 {
    fn foo() -> HashMap<String> {
        unimplemented!()
    }
}
```

File: src/main/kotlin/org/rust/ide/utils/import/RsImportHelper.kt [hunk 1]
```diff
@@ -23,9 +23,10 @@ object RsImportHelper {
         context: RsElement,
         elements: Collection<RsElement>,
         subst: Substitution = emptySubstitution,
-        useAliases: Boolean = false
+        useAliases: Boolean = false,
+        skipUnchangedDefaultTypeArguments: Boolean = false
```
[the remark this answers is not in the document]
Comment (mchernyavsky) [reply src/main/kotlin/org/rust/ide/utils/import/RsImportHelper.kt:27]: Maybe it make sense to change `skipUnchangedDefaultTypeArguments` to `true` for `Ty.renderInsertionSafe`, but not for `Ty.render`? And remove this parameter for `importTypeReferencesFromElements` / `getTypeReferencesInfoFromTys`?

[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/ide/utils/import/RsImportHelper.kt:27]: Probably. But it requires checking each feature that uses importing that may complicate this PR. But it's up to you

**Comment by mchernyavsky** [comment]
@Undin Fixed.

**Comment by Undin** [comment]
@mchernyavsky tests fail

**Comment by mchernyavsky** [comment]
@Undin fixed.

**Comment by mchernyavsky** [comment]
@Undin Rebased on master.

**Review by Undin — APPROVED** [review]
👍 
bors r+

**Pull request merged**